New module: monitoring/grafana_datasource

seuf opened the pull request:
##### SUMMARY
**grafana_datasource** is a new Ansible module that allow you to create/update/delete grafana datasources

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/modules/monitoring/grafana_datasource.py

##### ANSIBLE VERSION
```
ansible 2.3.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, May 10 2017, 20:04:28) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```

##### ADDITIONAL INFORMATION

This module can manage grafana datasource.
I've only tested it with influxdb and elasticsearch `ds_type`, but it should work with graphite, prometheus and opentsdb too.

Example : 
```
hosts: grafana
tasks:
  - name: creating influxdb datasource
    connection: local
    grafana_datasource:
      name: "influxdb_ansible"
      grafana_url: "http://grafana.company.com"
      grafana_user: "admin"
      grafana_password: "admin"
      org_id: "1"
      ds_type: "influxdb"
      url: "http://influxdb.company.com:8086"
      database: "telegraf"
      state: present
```

Commit:
Add monitoring/grafana_datasource module

added ds type mysql and postgres + check of parameters

nerzhul commented on lib/ansible/modules/monitoring/grafana_datasource.py:
```diff
@@ -0,0 +1,469 @@
+      - opentsdb time resolution (1 for second, 2 for millisecond)
+    default: 1
+  validate_certs:
+    required: false
```
missing default value

nerzhul commented on lib/ansible/modules/monitoring/grafana_datasource.py:
```diff
@@ -0,0 +1,469 @@
+            interval=dict(type='str'),
+            tsdb_version=dict(type='int', default=1, choices=[1, 2, 3]),
+            tsdb_resolution=dict(type='int', default=1, choices=[1, 2]),
+            validate_certs=dict(type='bool')
```
please set default to True

nerzhul reviewed (changes requested):
First review, in 1 min

gundalow commented on lib/ansible/modules/monitoring/grafana_datasource.py:
```diff
@@ -0,0 +1,469 @@
+    required: true
+    choices: [graphite,prometheus,elasticsearch,influxdb,opentsdb]
+    description:
+     - Type of the datasource C(graphite),C(prometheus),C(elasticsearch),C(influxdb),C(opentsdb)
```
No need to repeat the choices.

gundalow commented on lib/ansible/modules/monitoring/grafana_datasource.py:
```diff
@@ -0,0 +1,469 @@
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
```
Please update as per
https://github.com/ansible/ansible/pull/31930/files

```
required_together=[['grafana_user', 'grafana_password', 'org_id']],
mutually_exclusive=[['grafana_user', 'grafana_api_key']],
```

seuf replied:
I've also added `required_if` for some parameters..

gundalow commented on lib/ansible/modules/monitoring/grafana_datasource.py:
```diff
@@ -0,0 +1,469 @@
+    required: false
+    description:
+      - Name of the database for the datasource.
+      - This options is required when the ds_type is influxdb or elasticsearch (index name)
```
Should be enforced by argspec

seuf replied:
the `tls_ca_cert` is not mandatory when tls_client auth is enabled.
It's just an other option to trust the certificate of the datasource.

seuf replied:
done with `required_if`

gundalow commented on lib/ansible/modules/monitoring/grafana_datasource.py:
```diff
@@ -0,0 +1,469 @@
+    required: false
+    choices: [1, 2]
+    description:
+      - opentsdb time resolution (1 for second, 2 for millisecond)
```
Can we set the choices to `[seconds, milliseconds]` then convert it before passing the data to grafana?

seuf replied:
ok, I've changed the available choices then convert it for grafana api.

gundalow commented on lib/ansible/modules/monitoring/grafana_datasource.py:
```diff
@@ -0,0 +1,469 @@
+    default: timestamp
+    description:
+      - Name of the time field in elasticsearch ds
+      - For example @timestamp
```
`For example C(@timestamp).`

gundalow commented on lib/ansible/modules/monitoring/grafana_datasource.py:
```diff
@@ -0,0 +1,469 @@
+    required: false
+    description:
+      - tls ca certificate for self signed certificates.
+      - only used when tls_client_cert and tls_client_key are set
```
required_together?

seuf replied:
the tls_ca_cert is not mandatory when tls_client auth is enabled.
It's just an other option to trust the certificate of the datasource.

gundalow reviewed (changes requested):
Same comments as in https://github.com/ansible/ansible/pull/31930 apply here

gundalow commented:
Cut off for new functionality for 2.5 is 15th Jan.

seuf commented:
Ok. All requested changes are done except for tls_ca_cert option which is not mandatory.

Commit:
Added option max_concurrent_shard_requests for es_version >= 5.6

seuf commented:
OK, I've added the es_version 56 choices and the new max_conccurent_shard_requests option to limit the number of shards requests by elasticsearch.
I've fixed missing documentation, and formating with {} pattern (replaced with `"foo %s bar" % var`).
All checks have passed. 
We are currently using this module in production to automatically create grafana datasources for different organisations. it would be awesome to have it in ansible 2.5.

gundalow commented:
Merged into `devel` will be released in `2.5.0`
Thanks!
